### ISSUE TITLE
A little problem about class label

### ISSUE THREAD

**Issue by JiangtianPan** [CSAILVision/semantic-segmentation-pytorch#76]
Hi, thanks for your opensource. But several problems occurred during my training and testing.
1, I do wonder what the class index 0 stands for? I did not see any description about index 0 in the object info 150 text file.
2, when I use this model for training another dataset, which focuses on human parsing and all backgrounds are labeled as class index 0. But this model cannot understand it and parses the complex background as numerous part. I wonder if there is any problem I need to care about when dealing with this kind of dataset?

If you can reply to me, I would be extremely appreciated

**Comment by JiangtianPan** [comment]
@hangzhaomit

**Comment by hangzhaomit** [comment]
@JiangtianPan We use 0 as 'unlabeled' in the dataset, and ignore it during training.

**Comment by yangdonghan50** [comment]
> @JiangtianPan We use 0 as 'unlabeled' in the dataset, and ignore it during training.

Actually you ignore the label 0 by using ' label - 1 ' and then, nn.NLLLoss(ignore_index=-1), right?
I just wonder, in dataset.py :

batch_widths = np.zeros(self.batch_per_gpu, np.int32)
batch_heights = np.zeros(self.batch_per_gpu, np.int32)

img = self.img_transform(img)
segm = self.segm_transform(segm)

batch_images[i][:, :img.shape[1], :img.shape[2]] = img
batch_segms[i][:segm.shape[0], :segm.shape[1]] = segm

won't this method introduce additional label 0 (-1 ~ 149), which may affect the loss function? 
why not fill label -1 which is ignored here?

**Comment by yangdonghan50** [comment]
> @JiangtianPan We use 0 as 'unlabeled' in the dataset, and ignore it during training.

Actually you ignore the label 0 by using ' label - 1 ' and then, nn.NLLLoss(ignore_index=-1), right?
I just wonder, in dataset.py :

batch_widths = np.zeros(self.batch_per_gpu, np.int32)
batch_heights = np.zeros(self.batch_per_gpu, np.int32)

img = self.img_transform(img)
segm = self.segm_transform(segm)

batch_images[i][:, :img.shape[1], :img.shape[2]] = img
batch_segms[i][:segm.shape[0], :segm.shape[1]] = segm

won't this method introduce additional label 0 (-1 ~ 149), which may affect the loss function?
why not fill label -1 which is ignored here?

**Comment by RicoSuaveGuapo** [comment]
> Actually you ignore the label 0 by using ' label - 1 ' and then, nn.NLLLoss(ignore_index=-1), right?
not exactly, they did input the label as 0 (from seg .png), but while the `segm = torch.from_numpy(segm).long() - 1` (in the `segm_transform` of dataset.py) happens, the output (the final label) of the dataloader is -1.
That is why they use ignore = -1 in the loss
